File: PHDtools/html/Diagnosis-tools/primer-design.php
<?php
    $dtime = date('Y-m-d-h-i-s', time());
    $perl_path = "/home/dell/miniconda3/envs/Perl/bin/perl";
    $analysis_path = "/home/dell/data/Web-Server/Primer-design/".$dtime."/";
    $checkfa = "/var/www/other_scripts/bio-pipelines/Primer-design/Functions/checkfa.pl";
    $design_pipeline = "/var/www/other_scripts/bio-pipelines/Primer-design/primer-design.pl";
    $Result_store_path = "/home/dell/data/Web-Server/Result-store/Primer-design/".$dtime."/";
    if(empty($_POST['type'])){
        die('Please choose one of the PCR primer design type!<br>');
    }else{
	shell_exec("mkdir ".$analysis_path);
        shell_exec("mkdir ".$analysis_path.".../");
        shell_exec("chmod -R 777 ".$analysis_path);
        $save_file = fopen($analysis_path."target_seq.fasta", 'w+');
        fwrite($save_file, $_POST['target']);
        fclose($save_file);
        $command = $perl_path." ".$checkfa." -WorkPath ".$analysis_path." -target target_seq.fasta";
        $result = `$command`;
        $result = str_replace(array("\n", "\r"), "", $result);
        if(strcmp("TRUE", $result)==0){
            if($_FILES['conserve']['error'] == 0){
                $cons_tmp = $_FILES['conserve']['tmp_name'];
                $cons_rname = $_FILES['conserve']['name'];
                $cons_name = "conserve_info.tsv";
                if(move_uploaded_file($cons_tmp, $analysis_path.$cons_name)){
                    echo "$cons_rname file uploaded!<br>";
                    shell_exec("chmod -R 777 ".$analysis_path);
                    $command = $perl_path." ".$design_pipeline." -WorkPath ".$analysis_path." -target target_seq.fasta -type ".$_POST['type']." -consFILE ".$cons_name." -StorePath ".$Result_store_path;
                    shell_exec($command);
                    shell_exec("chmod 777 ".$analysis_path.".../shell.sh");
                    echo "Your Download code is: Primer-design_".$dtime."<br>";
                    echo "Please waiting for the analysis, remember the download code and you can close the window!<br>";
                    pclose(popen('bash '.$analysis_path.'.../shell.sh &', 'r'));
                }else{
                    shell_exec("rm -rf ".$analysis_path);
                    die('The conserve information file upload error!<br>');
                }
            }else{
                $command = $perl_path." ".$design_pipeline." -WorkPath ".$analysis_path." -target target_seq.fasta -type ".$_POST['type']." -StorePath ".$Result_store_path;
                shell_exec($command);
                shell_exec("chmod 777 ".$analysis_path.".../shell.sh");
                echo "Your Download code is: Primer-design_".$dtime."<br>";
                echo "Please waiting for the analysis, remember the download code and you can close the window!<br>";
                pclose(popen('bash '.$analysis_path.'.../shell.sh &', 'r'));
            }
        }else{
            shell_exec("rm -rf ".$analysis_path);
            die('Plsease input the FASTA format sequence!<br>');
        }
    }
?>

File: PHDtools/other_scripts/bio-pipelines/Primer-design/primer-design.pl
#!/usr/bin/perl -w

=head1 Description
	Primer design
Usage
	perl primer-design.pl -WorkPath <path> -target <target sequence file> -type <nPCR or qPCR>
	or
	perl primer-design.pl -WorkPath <path> -target <target sequence file> -type <nPCR or qPCR> -consFILE <conserved score file>
Parameters
	-WorkPath	[str]	Input the analysis path
	-target	[str]	Input the target nucleotide sequence but no more than 300 bp.
	-type	[str]	Input the PCR type (nPCR or qPCR)
	-consFILE	[str] Input the converse score file, output from conserved sequence determine
	-h/-help	[str] print help
Auther
	Dongyan Xiong
Edit Time
	2021.08.22 00:53 0.0.1
	2022.07.21 00:57 0.0.2
=cut

use strict;
use warnings;
use Bio::SeqIO;
use Getopt::Long;

my $perl = "/home/dell/miniconda3/envs/Perl/bin/perl";
my $primerdesign = "/var/www/other_scripts/bio-pipelines/Primer-design/Functions/PrimerDesign.pl";

my $analysis_path = "";
my ($target_file, $PCR_type, $cons_file, $help) = ("", "", "", "");
my $storage_path = "";
GetOptions(
	'WorkPath=s' => \$analysis_path,
	'target=s' => \$target_file,
	'type=s' => \$PCR_type,
	'consFILE:s' => \$cons_file,
	'StorePath=s' => \$storage_path,
	'h|help:s' => \$help,
);
die `pod2text $0` if((!$analysis_path) or (!$PCR_type) or (!$target_file) or (!$storage_path) or ($help));


my @shell;
push @shell, "#!/usr/bin/bash\n";
my $fasta = Bio::SeqIO -> new(-file => $analysis_path.$target_file, -format => 'fasta');
my $seq = $fasta -> next_seq;
my $sequence = $seq -> seq;
$sequence = uc($sequence);
die `pod2text $0` if(length($sequence) > 360);
my $checkseq = $sequence;
$checkseq =~ s/A//g;
$checkseq =~ s/T//g;
$checkseq =~ s/C//g;
$checkseq =~ s/G//g;
die `pod2text $0` if(length($checkseq) >= 1);
die `pod2text $0` if($PCR_type ne "qPCR" and $PCR_type ne "nPCR");

if(!$cons_file){
	push @shell, $perl." ".$primerdesign." -WorkPath ".$analysis_path." -sequence ".$sequence." -type ".$PCR_type."\n";
}
else{
	push @shell, $perl." ".$primerdesign." -WorkPath ".$analysis_path." -sequence ".$sequence." -type ".$PCR_type." -consFILE ".$cons_file."\n";
}
push @shell, "chmod -R 777 ".$analysis_path."\n";
push @shell, "mkdir ".$storage_path."\n";
push @shell, "chmod -R 777 ".$storage_path."\n";
push @shell, "mv ".$analysis_path."Primers.tsv ".$storage_path."\n";
push @shell, "chmod -R 777 ".$storage_path."\n";
push @shell, "rm -rf ".$analysis_path."\n";
open(OUT, ">", $analysis_path.".../shell.sh");
print OUT @shell;
close OUT;
